src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.Helpers.cs

```diff
@@ -0,0 +1,872 @@
+        /// first three bytes as a three-byte UTF-8 subsequence and returns the UTF-16 representation.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint ExtractCharFromFirstThreeByteSequence(uint value)
+        {
+            if (BitConverter.IsLittleEndian)
+            {
+                // value = [ ######## | 10xxxxxx 10yyyyyy 1110zzzz ]
+                return ((value & 0x003F0_000u) >> 16)
+                    | ((value & 0x0000_3F00u) >> 2)
+                    | ((value & 0x0000_000Fu) << 12);
+            }
+            else
+            {
+                // value = [ 1110zzzz 10yyyyyy 10xxxxxx | ######## ]
+                return ((value & 0x0F00_0000u) >> 12)
+                    | ((value & 0x003F_0000u) >> 10)
+                    | ((value & 0x0000_3F00u) >> 8);
+            }
+        }
+
+        /// <summary>
+        /// Given a machine-endian DWORD which four bytes of UTF-8 data, interprets the
+        /// first two bytes as a two-byte UTF-8 subsequence and returns the UTF-16 representation.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint ExtractCharFromFirstTwoByteSequence(uint value)
+        {
+            Debug.Assert(DWordBeginsWithUtf8TwoByteMask(value) && !DWordBeginsWithOverlongUtf8TwoByteSequence(value));
+
+            if (BitConverter.IsLittleEndian)
+            {
+                // value = [ ######## ######## | 10xxxxxx 110yyyyy ]
+                uint leadingByte = (uint)(byte)value << 6;
+                return (uint)(byte)(value >> 8) + leadingByte - (0xC0u << 6) - 0x80u; // remove header bits
+            }
+            else
+            {
+                // value = [ 110yyyyy 10xxxxxx | ######## ######## ]
+                return (char)(((value & 0x1F00_0000u) >> 18) | ((value & 0x003F_0000u) >> 16));
+            }
+        }
+
+        /// <summary>
+        /// Given a machine-endian DWORD which four bytes of UTF-8 data, interprets the input as a
+        /// four-byte UTF-8 sequence and returns the machine-endian DWORD of the UTF-16 representation.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+        private static uint ExtractCharsFromFourByteSequence(uint value)
+        {
+            if (BitConverter.IsLittleEndian)
+            {
+                if (Bmi2.IsSupported)
+                {
+                    // need to reverse endianness for bit manipulation to work correctly
+                    value = BinaryPrimitives.ReverseEndianness(value);
+
+                    // value = [ 11110uuu 10uuzzzz 10yyyyyy 10xxxxxx ]
+                    // want to return [ 110110wwwwxxxxxx 110111xxxxxxxxxx ]
+                    // where wwww = uuuuu - 1
+
+                    uint highSurrogateChar = Bmi2.ParallelBitExtract(value, 0b00000111_00111111_00110000_00000000u);
```

Comment:
Isn't this really slow on some CPUs?

Reply:
Possibly. Agner's instruction tables say that Ryzen in particular has high latency with this instruction. Intel systems do seem to benefit significantly from this, however. Since Ryzen is primarily found in consumer environments I think it makes sense to optimize for the non-Ryzen case. (Additionally, this code path is only hit when processing non-BMP characters, which for now is primarily emoji. It's not a common code path.)

Reply:
> Since Ryzen is primarily found in consumer environments

Fair enough. But we might want to be careful about such assumptions. Azure was the Lv2-Series which are powered by EPYC (the Ryzen Server chips) and I believe other cloud providers have similar offerings.

Reply:
Aside: can the JIT determine if Intel / Amd on spit out optimal code for each in this case?

Reply:
Yup, it's definitely a trade-off. Since it's just an implementation detail with no end-user visible behavioral change we can revisit this as the ecosystem evolves. (Or, if anybody feels strongly that the trade-off isn't worth it, we could address it sooner.) For now it does improve the benchmarks in the systems we have, and we hope our systems are representative of the typical customer's or cloud provider's systems. :)

@gfoidl JIT should have that knowledge since it's already performing a cpuid check to figure out supported ISAs, but I don't think we'd want to expose `IsIntel` or `IsAmd` via our public surface area.

Reply:
> want to expose `IsIntel` or `IsAmd`

I don't won't this either, as it complicates things quite a lot (less readable code, more code-path, with the potential of hitting JIT inline-limits, ...). 
But JIT could do this internally. Like in this case use `Bmi2.ParallelBitExtract` on Intel, and on Ryzen the there faster codegen.

Maybe something to revisit in the future (if there's any real concern about that).

Reply:
I guess the open question in this case is what is the "faster codegen"? I can't imagine the JIT wants to emulate the `pext` instruction in software (since we don't do that with any other intrinsics), so would `Bmi2.IsSupported` just return _false_? But Ryzen doesn't have a problem with other BMI2 instructions, so isn't this needlessly punishing them? It seems the C# code is really the only place that has good context as to what an appropriate fallback mechanism might be.

These are rherotical questions. I don't expect anybody to have good answers to them, instead just giving a sense of how deep the rabbit hole goes.